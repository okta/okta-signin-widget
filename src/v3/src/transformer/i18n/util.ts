/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
