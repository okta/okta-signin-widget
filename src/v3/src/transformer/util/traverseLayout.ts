/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
