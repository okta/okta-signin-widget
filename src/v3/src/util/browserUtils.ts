/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
