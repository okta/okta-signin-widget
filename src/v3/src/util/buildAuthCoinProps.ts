/*
 * Copyright (c) 2025-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  AUTHENTICATOR_KEY, DEVICE_ENROLLMENT_TYPE, EMAIL_AUTHENTICATOR_TERMINAL_KEYS, IDX_STEP,
} from '../constants';
import { AuthCoinProps } from '../types';
import { getAuthenticatorKey } from './getAuthenticatorKey';
import { containsOneOfMessageKeys } from './idx';

export const buildAuthCoinProps = (
  transaction?: IdxTransaction | null,
): AuthCoinProps | undefined => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.CONSENT_EMAIL_CHALLENGE) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (nextStep?.name === IDX_STEP.PIV_IDP) {
    return { authenticatorKey: IDX_STEP.PIV_IDP };
  }

  if (nextStep?.name === IDX_STEP.REDIRECT_IDVERIFY && nextStep?.idp?.id) {
    return { authenticatorKey: nextStep.idp.id };
  }

  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  if (transaction.context?.deviceEnrollment?.value?.name === DEVICE_ENROLLMENT_TYPE.ODA) {
    return { authenticatorKey: AUTHENTICATOR_KEY.OV };
  }

  const authenticatorKey = getAuthenticatorKey(transaction);
  if (authenticatorKey) {
    return {
      authenticatorKey,
      // @ts-ignore logoUri missing from interface
      url: nextStep?.relatesTo?.value.logoUri,
    };
  }

  return undefined;
};
