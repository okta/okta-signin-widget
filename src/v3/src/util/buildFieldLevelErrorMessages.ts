/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { WidgetMessage } from '../types';

type FieldLevelErrorMessages = {
  errorMessage?: string;
  errorMessageList?: string[];
};

export const buildFieldLevelErrorMessages = (
  errors: WidgetMessage[] | undefined,
) : FieldLevelErrorMessages => {
  const hasErrors = typeof errors !== 'undefined' && errors.length > 0;

  if (hasErrors) {
    if (errors.length === 1) {
      const error = errors[0];
      // Covers nested messages, e.g. password requirements not met
      // The SIW does not have messages nested more than once
      if (Array.isArray(error.message)) {
        return {
          errorMessage: error.description,
          errorMessageList: error.message
            .filter((msg) => typeof msg.message === 'string')
            .map((msg) => msg.message as string),
        };
      }
      return { errorMessage: error.message };
    }
    return {
      errorMessageList: errors
        .filter((error) => typeof error.message === 'string')
        .map((error) => error.message as string),
    };
  }

  return {};
};
