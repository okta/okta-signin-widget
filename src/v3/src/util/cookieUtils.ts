/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import Cookies from 'js-cookie';

const DEFAULT_EXPIRY_DAYS = 365;
const DEFAULT_COOKIE_PATH = '/';
const LAST_USERNAME_COOKIE_NAME = 'ln';
const USER_AUTHENTICATED_COOKIE_NAME = 'isuauth';

export const setCookieString = (
  name: string,
  value: string,
  expiryDays?: number,
  path?: string,
): void => {
  Cookies.set(
    name,
    value,
    { expires: expiryDays ?? DEFAULT_EXPIRY_DAYS, path: path ?? DEFAULT_COOKIE_PATH },
  );
};

export const getCookieString = (name: string): string => Cookies.get(name) as string;

export const getCookieUserAuthenticated = (): string => getCookieString(
  USER_AUTHENTICATED_COOKIE_NAME,
);

export const removeCookie = (name: string): void => Cookies.remove(name);

export const setUsernameCookie = (username: string): void => setCookieString(
  LAST_USERNAME_COOKIE_NAME,
  username,
);

export const setCookieUserAuthenticated = (): void => {
  Cookies.set(
    USER_AUTHENTICATED_COOKIE_NAME,
    '1',
    { expires: 1, path: DEFAULT_COOKIE_PATH },
  );
};

export const getUsernameCookie = (): string => getCookieString(LAST_USERNAME_COOKIE_NAME);

export const removeUsernameCookie = (): void => removeCookie(LAST_USERNAME_COOKIE_NAME);

export const removeUserAuthenticatedCookie = (): void => removeCookie(
  USER_AUTHENTICATED_COOKIE_NAME,
);
