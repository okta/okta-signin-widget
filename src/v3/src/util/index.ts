/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
