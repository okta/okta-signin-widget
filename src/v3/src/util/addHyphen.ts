/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
