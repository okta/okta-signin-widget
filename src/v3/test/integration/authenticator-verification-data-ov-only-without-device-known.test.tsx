/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import mockResponse from '../../../../playground/mocks/data/idp/idx/authenticator-verification-data-ov-only-without-device-known.json';
import { createAuthJsPayloadArgs, setup } from './util';

describe('authenticator-verification-data-ov-only-without-device-known', () => {
  it('should render form', async () => {
    const { container, findByRole } = await setup({ mockResponse });
    const heading = await findByRole('heading', { level: 2 });
    expect(heading.textContent).toBe('Verify it\'s you with a security method');
    expect(container).toMatchSnapshot();
  });

  it('should send correct payload when clicking "Get a push notification"', async () => {
    const {
      authClient, user, findByRole,
    } = await setup({ mockResponse, widgetOptions: { features: { autoFocus: true } } });

    const heading = await findByRole('heading', { level: 2 });
    expect(heading.textContent).toBe('Verify it\'s you with a security method');

    const authenticatorButton = await findByRole('button', { name: 'Get a push notification' });
    await user.click(authenticatorButton);
    expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
      ...createAuthJsPayloadArgs('POST', 'idp/idx/challenge', {
        authenticator: { id: 'aut13qrZReYpIib7R0g4', methodType: 'push' },
      }, 'application/ion+json; okta-version=1.0.0'),
    );
  });
});
