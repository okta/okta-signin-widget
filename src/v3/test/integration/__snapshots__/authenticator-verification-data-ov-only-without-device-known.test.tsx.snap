// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticator-verification-data-ov-only-without-device-known should render form 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <div
      class="MuiScopedCssBaseline-root emotion-1"
    >
      <main
        class="auth-container main-container MuiBox-root emotion-2"
        data-commit="b9bbc0140703c3fbf0e2e58920362e70"
        data-version="0.0.0"
        dir="ltr"
        id="okta-sign-in"
        lang="en"
      >
        <div
          class="MuiScopedCssBaseline-root emotion-3"
        >
          <div
            class="MuiBox-root emotion-4"
          >
            <div
              class="MuiBox-root emotion-5"
            >
              <div
                class="okta-sign-in-header auth-header authCoinSpacing MuiBox-root emotion-6"
              >
                <h1
                  class="MuiTypography-root MuiTypography-h1 emotion-7"
                />
                <div
                  aria-hidden="true"
                  class="mfa-okta-verify MuiBox-root emotion-8"
                  data-se="factor-beacon"
                >
                  <svg
                    aria-labelledby="mfa-okta-verify"
                    fill="none"
                    height="48"
                    role="img"
                    viewBox="0 0 48 48"
                    width="48"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <title
                      id="mfa-okta-verify"
                    >
                      Okta Verify
                    </title>
                    <circle
                      class="siwFillBg"
                      cx="24"
                      cy="24"
                      fill="#F5F5F6"
                      r="24"
                    />
                    <path
                      clip-rule="evenodd"
                      d="M32.872 22.853a8.947 8.947 0 1 1-2.288-4.91L24 24.498l-3.205-3.19c-.482-.48-1.322-.48-1.804 0a1.26 1.26 0 0 0-.374.898c0 .34.133.659.374.898l4.107 4.09c.24.239.561.371.902.371.336 0 .664-.136.902-.372L38.636 13.52a18.089 18.089 0 0 0-1.634-1.967A17.947 17.947 0 0 0 24 6C14.059 6 6 14.059 6 24s8.059 18 18 18 18-8.059 18-18c0-2.972-.722-5.775-1.998-8.245l-7.13 7.098Z"
                      fill="#00297A"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-9"
              >
                <form
                  class="o-form MuiBox-root emotion-10"
                  data-se="o-form"
                  novalidate=""
                >
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="identifier-container MuiBox-root emotion-13"
                      >
                        <div
                          class="MuiBox-root emotion-14"
                        >
                          <span
                            class="MuiBox-root emotion-15"
                          >
                            <svg
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-16"
                              fill="none"
                              focusable="false"
                              role="img"
                              viewBox="0 0 24 24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                clip-rule="evenodd"
                                d="M15.5 6.5a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Zm2 0a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0ZM5 23c0-2.357.694-4.363 1.886-5.762C8.064 15.856 9.782 15 12 15c2.215 0 3.934.862 5.113 2.25C18.307 18.652 19 20.66 19 23h2c0-2.722-.807-5.216-2.363-7.046C17.067 14.107 14.785 13 12 13c-2.782 0-5.064 1.097-6.636 2.941C3.806 17.77 3 20.263 3 23h2Z"
                                fill="currentColor"
                                fill-rule="evenodd"
                              />
                              <title>
                                User
                              </title>
                            </svg>
                          </span>
                          <span
                            aria-label="testUser@okta.com"
                            class="identifier no-translate MuiBox-root emotion-17"
                            data-se="identifier"
                            title="testUser@okta.com"
                          >
                            testUser@okta.com
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-19"
                      >
                        <h2
                          class="MuiTypography-root MuiTypography-h4 emotion-20"
                          data-se="o-form-head"
                          id="authenticator-verification-data_Title_Verify_its_you_with_a_security_method_okta_verify_aut13qrZReYpIib7R0g4_1"
                          tabindex="-1"
                        >
                          Verify it's you with a security method
                        </h2>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-2"
                    >
                      <div
                        class="MuiBox-root emotion-19"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-23"
                          data-se="o-form-explain"
                          id="authenticator-verification-data_Description_Select_from_the_following_options_okta_verify_aut13qrZReYpIib7R0g4_2"
                        >
                          Select from the following options
                        </p>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <ul
                        class="MuiBox-root emotion-25"
                        data-se="authenticator-verify-list"
                      >
                        <li
                          class="MuiBox-root emotion-26"
                        >
                          <button
                            aria-describedby="okta_verify_0-description okta_verify_0-ctaLabel"
                            aria-labelledby="okta_verify_0-label"
                            class="authenticator-row MuiBox-root emotion-27"
                            data-se="authenticator-button"
                            tabindex="0"
                            type="button"
                          >
                            <div
                              class="authenticator-icon-container MuiBox-root emotion-2"
                              data-se="authenticator-icon"
                            >
                              <div
                                aria-hidden="true"
                                class="mfa-okta-verify authenticator-icon MuiBox-root emotion-29"
                                data-se="factor-beacon"
                              >
                                <svg
                                  aria-labelledby="okta_verify_0"
                                  fill="none"
                                  height="48"
                                  role="img"
                                  viewBox="0 0 48 48"
                                  width="48"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <title
                                    id="okta_verify_0"
                                  >
                                    Okta Verify
                                  </title>
                                  <circle
                                    class="siwFillBg"
                                    cx="24"
                                    cy="24"
                                    fill="#F5F5F6"
                                    r="24"
                                  />
                                  <path
                                    clip-rule="evenodd"
                                    d="M32.872 22.853a8.947 8.947 0 1 1-2.288-4.91L24 24.498l-3.205-3.19c-.482-.48-1.322-.48-1.804 0a1.26 1.26 0 0 0-.374.898c0 .34.133.659.374.898l4.107 4.09c.24.239.561.371.902.371.336 0 .664-.136.902-.372L38.636 13.52a18.089 18.089 0 0 0-1.634-1.967A17.947 17.947 0 0 0 24 6C14.059 6 6 14.059 6 24s8.059 18 18 18 18-8.059 18-18c0-2.972-.722-5.775-1.998-8.245l-7.13 7.098Z"
                                    fill="#00297A"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </div>
                            </div>
                            <div
                              class="authenticator-description MuiBox-root emotion-30"
                            >
                              <h3
                                class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-31"
                                data-se="authenticator-button-label"
                                id="okta_verify_0-label"
                              >
                                Get a push notification
                              </h3>
                              <p
                                class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph authenticator-description--text emotion-32"
                                data-se="authenticator-button-description"
                                id="okta_verify_0-description"
                              >
                                Okta Verify
                              </p>
                              <div
                                class="cta-button authenticator-button MuiBox-root emotion-33"
                                data-se="okta_verify-push"
                              >
                                <span
                                  class="button select-factor link-button MuiBox-root emotion-34"
                                  data-se="cta-button-label"
                                  id="okta_verify_0-ctaLabel"
                                >
                                  Select
                                </span>
                                <svg
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-35"
                                  fill="none"
                                  focusable="false"
                                  role="img"
                                  viewBox="0 0 24 24"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="m18.354 10.94-4.647-4.647-1.414 1.414L15.586 11H5v2h10.586l-3.293 3.293 1.414 1.414 4.647-4.646a1.5 1.5 0 0 0 0-2.122Z"
                                    fill="currentColor"
                                    fill-rule="evenodd"
                                  />
                                  <title>
                                    Select
                                  </title>
                                </svg>
                              </div>
                            </div>
                          </button>
                        </li>
                        <li
                          class="MuiBox-root emotion-26"
                        >
                          <button
                            aria-describedby="okta_verify_1-description okta_verify_1-ctaLabel"
                            aria-labelledby="okta_verify_1-label"
                            class="authenticator-row MuiBox-root emotion-27"
                            data-se="authenticator-button"
                            tabindex="0"
                            type="button"
                          >
                            <div
                              class="authenticator-icon-container MuiBox-root emotion-2"
                              data-se="authenticator-icon"
                            >
                              <div
                                aria-hidden="true"
                                class="mfa-okta-verify authenticator-icon MuiBox-root emotion-29"
                                data-se="factor-beacon"
                              >
                                <svg
                                  aria-labelledby="okta_verify_1"
                                  fill="none"
                                  height="48"
                                  role="img"
                                  viewBox="0 0 48 48"
                                  width="48"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <title
                                    id="okta_verify_1"
                                  >
                                    Okta Verify
                                  </title>
                                  <circle
                                    class="siwFillBg"
                                    cx="24"
                                    cy="24"
                                    fill="#F5F5F6"
                                    r="24"
                                  />
                                  <path
                                    clip-rule="evenodd"
                                    d="M32.872 22.853a8.947 8.947 0 1 1-2.288-4.91L24 24.498l-3.205-3.19c-.482-.48-1.322-.48-1.804 0a1.26 1.26 0 0 0-.374.898c0 .34.133.659.374.898l4.107 4.09c.24.239.561.371.902.371.336 0 .664-.136.902-.372L38.636 13.52a18.089 18.089 0 0 0-1.634-1.967A17.947 17.947 0 0 0 24 6C14.059 6 6 14.059 6 24s8.059 18 18 18 18-8.059 18-18c0-2.972-.722-5.775-1.998-8.245l-7.13 7.098Z"
                                    fill="#00297A"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </div>
                            </div>
                            <div
                              class="authenticator-description MuiBox-root emotion-30"
                            >
                              <h3
                                class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-31"
                                data-se="authenticator-button-label"
                                id="okta_verify_1-label"
                              >
                                Enter a code
                              </h3>
                              <p
                                class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph authenticator-description--text emotion-32"
                                data-se="authenticator-button-description"
                                id="okta_verify_1-description"
                              >
                                Okta Verify
                              </p>
                              <div
                                class="cta-button authenticator-button MuiBox-root emotion-33"
                                data-se="okta_verify-totp"
                              >
                                <span
                                  class="button select-factor link-button MuiBox-root emotion-34"
                                  data-se="cta-button-label"
                                  id="okta_verify_1-ctaLabel"
                                >
                                  Select
                                </span>
                                <svg
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-35"
                                  fill="none"
                                  focusable="false"
                                  role="img"
                                  viewBox="0 0 24 24"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="m18.354 10.94-4.647-4.647-1.414 1.414L15.586 11H5v2h10.586l-3.293 3.293 1.414 1.414 4.647-4.646a1.5 1.5 0 0 0 0-2.122Z"
                                    fill="currentColor"
                                    fill-rule="evenodd"
                                  />
                                  <title>
                                    Select
                                  </title>
                                </svg>
                              </div>
                            </div>
                          </button>
                        </li>
                        <li
                          class="MuiBox-root emotion-26"
                        >
                          <button
                            aria-describedby="okta_verify_2-description okta_verify_2-ctaLabel"
                            aria-labelledby="okta_verify_2-label"
                            class="authenticator-row MuiBox-root emotion-27"
                            data-se="authenticator-button"
                            tabindex="0"
                            type="button"
                          >
                            <div
                              class="authenticator-icon-container MuiBox-root emotion-2"
                              data-se="authenticator-icon"
                            >
                              <div
                                aria-hidden="true"
                                class="mfa-okta-verify authenticator-icon MuiBox-root emotion-29"
                                data-se="factor-beacon"
                              >
                                <svg
                                  aria-labelledby="okta_verify_2"
                                  fill="none"
                                  height="48"
                                  role="img"
                                  viewBox="0 0 48 48"
                                  width="48"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <title
                                    id="okta_verify_2"
                                  >
                                    Okta Verify
                                  </title>
                                  <circle
                                    class="siwFillBg"
                                    cx="24"
                                    cy="24"
                                    fill="#F5F5F6"
                                    r="24"
                                  />
                                  <path
                                    clip-rule="evenodd"
                                    d="M32.872 22.853a8.947 8.947 0 1 1-2.288-4.91L24 24.498l-3.205-3.19c-.482-.48-1.322-.48-1.804 0a1.26 1.26 0 0 0-.374.898c0 .34.133.659.374.898l4.107 4.09c.24.239.561.371.902.371.336 0 .664-.136.902-.372L38.636 13.52a18.089 18.089 0 0 0-1.634-1.967A17.947 17.947 0 0 0 24 6C14.059 6 6 14.059 6 24s8.059 18 18 18 18-8.059 18-18c0-2.972-.722-5.775-1.998-8.245l-7.13 7.098Z"
                                    fill="#00297A"
                                    fill-rule="evenodd"
                                  />
                                </svg>
                              </div>
                            </div>
                            <div
                              class="authenticator-description MuiBox-root emotion-30"
                            >
                              <h3
                                class="MuiTypography-root MuiTypography-h3 authenticator-label no-translate emotion-31"
                                data-se="authenticator-button-label"
                                id="okta_verify_2-label"
                              >
                                Use Okta FastPass
                              </h3>
                              <p
                                class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph authenticator-description--text emotion-32"
                                data-se="authenticator-button-description"
                                id="okta_verify_2-description"
                              >
                                Okta Verify
                              </p>
                              <div
                                class="cta-button authenticator-button MuiBox-root emotion-33"
                                data-se="okta_verify-signed_nonce"
                              >
                                <span
                                  class="button select-factor link-button MuiBox-root emotion-34"
                                  data-se="cta-button-label"
                                  id="okta_verify_2-ctaLabel"
                                >
                                  Select
                                </span>
                                <svg
                                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-35"
                                  fill="none"
                                  focusable="false"
                                  role="img"
                                  viewBox="0 0 24 24"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    clip-rule="evenodd"
                                    d="m18.354 10.94-4.647-4.647-1.414 1.414L15.586 11H5v2h10.586l-3.293 3.293 1.414 1.414 4.647-4.646a1.5 1.5 0 0 0 0-2.122Z"
                                    fill="currentColor"
                                    fill-rule="evenodd"
                                  />
                                  <title>
                                    Select
                                  </title>
                                </svg>
                              </div>
                            </div>
                          </button>
                        </li>
                      </ul>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <button
                        class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways MuiLink-button emotion-57"
                        data-se="cancel"
                        role="link"
                        type="button"
                      >
                        Back to sign in
                      </button>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
